central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -242,6 +251,9 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
 		}
 	}
 
+	ds.mutex.Lock()
```

Comment:
It's a bit of a misnomer, but I would say the following. We should be able to upsert many plops from different clusters at the same time. So perhaps this should be RLock and only when we need to delete we lock?

Reply:
This could really slow down upserts because now all upserts from all clusters are going to be serial

Reply:
`UpsertMany` in `pkg/search/postgres/store.go` has the following comment

```
                // Lock since copyFrom requires a delete first before being executed.  If multiple processes are updating
                // same subset of rows, both deletes could occur before the copyFrom resulting in unique constraint
                // violations
```
This seem like something that should be improved, in another PR. The point being that `UpsertMany` does use a delete. So, what kind of lock should be used here?

To decrease the time during which the lock is applied, I split `plopObjects` into two arrays. `newPlopObjects` are obviously new, so I don't think a lock has to be used when inserting them. `updatePlopObjects` already exist and need to be updated. The lock is only applied when updating them. It doesn't completely solve the problem, but should help a bit.

I did scale tests with 10 clusters. I didn't check the metrics for how long it takes to add PLOPs. I can try doing that.  Are there any other tests that you would like me to run to see if this could cause a problem?

Reply:
One thing that just occurred to me is that the UpsertMany function already uses a lock, so I don't think this will change the upserts from being parallel to serial. They are already serial.

Reply:
👍